libraries/tickv/SPEC.md

```diff
@@ -0,0 +1,510 @@
+header and check sum. The maximum length of the entire object is
+4KiB (0xFFF) or the region size, whichever is smaller.
+
+The `hashed_key` field stores the 64-bit (8 byte) output of the key hash.
+
+ObjectHeader is internal to TicKV and users of TicKV do not need to
+understand it.
+
+#### Object Value
+
+The Value component of the TicKV object is the value that the user wants to
+store.
+
+The values can be any length as long as they follow both:
+ * Don't span multiple regions. That limits the maximum value length to
+   `region_size - size_of::<ObjectHeader>()`
+ * Don't have a maximum length greater then 4KiB (0xFFF).
+
+#### Checksum
+
+The checksum is a hash of the entire object (not including the checksum).
+The checksum is calculated using the same hash algorithm used for the keys.
+
+A simple CRC isn't used to avoid depending on external crates and to
+re-use the already implemented Hash. This will potentially allow
+hardware accelerated implementation of `core::hash::Hasher` to be used
+for the checksum.
+
+### Object overhead
+
+Currently the overhead of an TicKV object is 21 bytes. Most of this is the 8
+bytes for the key hash and 8 bytes for a checksum.
+
+### Location of objects
+
+The region where a TicKV object is stored is dependent on the output of the
+key hash and the number of regions.
+
+TicKV determines the region an object will be stored or retrieved from using
+region numbers, starting from 0. The region number of an object is equal to the
+last two bytes of an object hashed-key modulo the number of regions.
+
+This will produce a number that is between zero and the total number of
+TicKV regions. This number is the regions number where the data is stored
+or loaded from.
+
+This allows us to quickly determine which flash region a key is stored in. This
+should improve lookup time by reducing the number of reads required.
+
+If a give region is full, we will start to look in neighboring regions. We first
+perform the same search in the next (increment region number) region. If this
+region doesn't exist (we are at the end of the regions) or if the region is full
+we then search in the previous region (decrement region number).
+
+When storing a object this process continues until we either:
+ * Search all regions
+ * Find a free space
+
+When retrieving an object the process continues until we either:
+ * Search all regions
+ * Find the key we are looking for
+ * Find a region that is empty
+
+### Invalidating keys
+
+Flash has the characteristic that although read/writes can happen at small
+granularities an erase operation requires a entire block (specified
+by the region size).
+
+Due to this removing a key with the `invalidate_key()` function does NOT
+remove anything from flash. Instead calling `invalidate_key()`
+will mark the `valid` boolean for that object as `false` (0).
+
+This is done because changing a `1` to a `0` in flash can be done with a
```

Comment:
Many chips allow a limited number of writes per erase cycle on a page. E.g., the EFMPG1 Reference Manual reads:

"It is possible to write words twice between each erase by keeping at 1 the bits that are not to be changed. Let us take as an example
writing two 16 bit values, 0xAAAA and 0x5555. To safely write them in the same flash word this method can be used:
• Write 0xFFFFAAAA (word in flash becomes 0xFFFFAAAA)
• Write 0x5555FFFF (word in flash becomes 0x5555AAAA)
Note:
• There is a maximum of two writes to the same word between each erase due to a physical limitation of the flash.
• During a write or erase, flash read accesses will be stalled, effectively halting code execution from flash. Code execution continues
upon write/erase completion. Code residing in RAM may be executed during a write/erase operation."

It's OK to say that TicKV only works on flash devices that allow N>K writes to a word. It looks like TicKV requires 2 writes be allowed. That means almost all (but perhaps not all) can work. You want this limitation up at the top, not at the bottom.

What happens if you re-insert a key? Does it fail or do you replace?

Reply:
Yeah, I have seen some devices with limitations.

As long as we can write twice between erases. Once to store the data and once to invalidate then it should work. From what I can tell it's not very common that you can't write that many times.

I have added a note to the README.

To re-insert a key you first have to delete the key then append it again. It won't replace the original key, it will be appended. The first key (that was invalidated) is marked for deletion and can't be read.